draft-ietf-tls-esni.md

```diff
@@ -822,22 +822,27 @@ error code.
 
 ### Handling HelloRetryRequest {#client-hrr}
 
-When the server sends a HelloRetryRequest, the client determines if ECH was
-accepted by checking the message for an "encrypted_client_hello" extension with
-an 8-byte payload equal to `hrr_accept_confirmation` as defined in
-{{backend-server}}. If found, the client presumes acceptance and handles the
-HelloRetryRequest using ClientHelloInner. Otherwise, it presumes rejection and
-handles the HelloRetryRequest using ClientHelloOuter. Note that the
-client-facing server does not send "encrypted_client_hello" in case of
+When the server sends a HelloRetryRequest, the client checks for the presence of
+an "encrypted_client_hello" extension. If none is found, then the client
+presumes rejection and handles the HelloRetryRequest using ClientHelloOuter.
+(Note that the client-facing server does not send this extension in its
+HelloRetryRequest.) Otherwise it proceeds as follows.
+
+If the extension's payload has a length other than 8, then the client aborts the
+handshake with an "illegal_parameter" alert. If the payload length is equal to
```

Comment:
How about we capture this in syntax, as `opaque HelloRetryRequestECH[8]` or whatever? I think syntax errors are already implicitly fatal and come with an alert (`decode_error`). Less special-case bits.

Looking at RFC8446, we seem to have a couple styles here. Sometimes we just define separate types for each context and link them in prose: `KeyShareClientHello`, `KeyShareHelloRetryRequest`, `KeyShareServerHello`. Sometimes we define one structure with a select:

```
      struct {
          select (Handshake.msg_type) {
              case new_session_ticket:   uint32 max_early_data_size;
              case client_hello:         Empty;
              case encrypted_extensions: Empty;
          };
      } EarlyDataIndication;

      struct {
          select (Handshake.msg_type) {
              case client_hello:
                   ProtocolVersion versions<2..254>;

              case server_hello: /* and HelloRetryRequest */
                   ProtocolVersion selected_version;
          };
      } SupportedVersions;
```

I guess the latter is a little interesting because, for HelloRetryRequest, `Handshake.msg_type` is still `server_hello`. Maybe that's why we did key shares in prose...

(Interestingly, our naming scheme for ECH is funny because it doesn't actually tell you the message, only who sends it. We say ServerECH rather than ECHEncryptedExtensions. Maybe time to rename since the server sends two of them now...)

Reply:
> I guess the latter is a little interesting because, for HelloRetryRequest, Handshake.msg_type is still server_hello. Maybe that's why we did key shares in prose...

Ah yes, that's exactly why key share doesn't do this. So I guess the single struct with `select` is the convention, but if you need to distinguish SH and HRR, as we do, do it in prose:
https://github.com/tlswg/tls13-spec/commit/9a9c3295140bf4880ca031a6512e350da108713b

Reply:
Something like this?
```
      struct {
          select (Handshake.msg_type) {
              case server_hello: // Well, HRR, but this isn't a codepoint AFAIK
                   opaque hrr_ech_accept_confirmation [8];
              case encrypted_extensions:
                   ECHConfigList retry_configs;
          };
      } ServerECH;
```

Reply:
Using a mix of the two styles seems a little odd. Normally, when extensions use a `select (Handshake.msg_type)`, we put everything in it, so the ClientHello extension would go here too. Though I suppose we don't have a SH extension so HRR using the SH code point does work. But I was thinking we just go the key_share route and define three structures:

```
    struct {
       ClientECHType type;
       select (ClientECH.type) {
           case outer:
               HpkeSymmetricCipherSuite cipher_suite;
               uint8 config_id;
               opaque enc<0..2^16-1>;
               opaque payload<1..2^16-1>;
           case inner:
               Empty;
       };
    } ECHClientHello;

    struct {
       opaque confirmation[8];
    } ECHHelloRetryRequest;
    // Or, equivalently, opaque ECHHelloRetryRequest[8], but the struct might be easier to read.

    struct {
       ECHConfigList retry_configs;
    } ECHEncryptedExtensions;
```

Renaming things might be too much churn, I dunno. That's just the convention RFC8446 used for key share.

Reply:
This looks good to me. @chris-wood, what do you think?

Reply:
Seems reasonable to me!

Reply:
Done.

Reply:
```suggestion
handshake with an "decode_error" alert. If the payload length is equal to
```

Reply:
(@cjpatton did you accidentally resolve this?)